frontend/src/main/scala/bloop/dap/DebugSession.scala

```diff
@@ -0,0 +1,182 @@
+package bloop.dap
+
+import java.net.{InetSocketAddress, Socket}
+import java.util.concurrent.TimeUnit
+
+import monix.execution.atomic.Atomic
+import com.microsoft.java.debug.core.adapter.{ProtocolServer => DapServer}
+import com.microsoft.java.debug.core.protocol.Messages.{Request, Response}
+import com.microsoft.java.debug.core.protocol.Requests._
+import com.microsoft.java.debug.core.protocol.{Events, JsonUtils}
+import monix.eval.Task
+import monix.execution.{Cancelable, CancelableFuture, Scheduler}
+
+import scala.collection.mutable
+import scala.concurrent.Promise
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Try
+import bloop.logging.Logger
+
+/**
+ *  This debug adapter maintains the lifecycle of the debuggee in separation from JDI.
+ *  The debuggee is started/closed together with the session.
+ *
+ *  This approach makes it necessary to handle the "launch" requests as the "attach" ones.
+ *  The JDI address of the debuggee is obtained through the [[DebugSessionLogger]]
+ */
+final class DebugSession(
```

Comment:
I really like the `DebugSession` and `DebugServer` abstractions and their implementation, good work 👍